torch/csrc/jit/script/exit_transforms.cpp

```diff
@@ -419,12 +430,25 @@ struct ExitTransformer {
     // exit values. since the exit does not extend outside this block,
     // update returned exit to false. then, reset the target_block to whatever
     // it was previously
-    // this will occur with a Loop block when transforming LoopContinuations
-    // and a Graph/Closure block when transforming ReturnStmts.
     if (target_block_ == block) {
-      TORCH_INTERNAL_ASSERT(getExitStatus(exit_pair) == ExitStatus::WILL);
-      registerBlockOutputs(block, exit_pair.exitValues());
-      exit_pair = ExitPair(false_val_, std::vector<Value*>({}));
+      if (getExitStatus(exit_pair) == ExitStatus::MIGHT) {
+        auto new_if =
+            graph_->create(prim::If, 0)->insertBefore(block->return_node());
+        new_if->addBlock();
+        new_if->addBlock();
+        new_if->addInput(exit_pair.hasExited());
+        addIfOutputs(new_if, exit_pair.exitValues(), block->outputs());
+        exit_pair = constructWillExitPair(new_if->outputs());
+      }
+
+      if (getExitStatus(exit_pair) == ExitStatus::WILL) {
+        removeOutputs(block);
+        registerBlockOutputs(block, exit_pair.exitValues());
+      }
+
+      // reset the exiting status so that e.g. an exception from a closure block
+      // wont propagate to the enclosing function
```

Comment:
Why don't we want to propagate exceptions up? (Sorry if you've explained to me and I forgot - if that's the case, then it's a good indicator that it deserves a comment :) )

Reply:
added comment inline